### PR TITLE
upgrade clickhouse jdbc driver 0.3.2-patch10

### PR DESCRIPTION
**Pull request by mzitnik**
Upgrade clickhouse jdbc driver to 0.3.2-patch10
@methodmissing 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -499,6 +499,7 @@ class BeamModulePlugin implements Plugin<Project> {
     def spark3_version = "3.1.2"
     def spotbugs_version = "4.0.6"
     def testcontainers_version = "1.16.3"
+    def testcontainers_version_clickhouse = "1.17.3"
```
Comment (johnjcasey) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:502]: What is the reason to not just update the testcontainers version globally?

Comment (mzitnik) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:502]: @johnjcasey  Yes i can, just was not sure that it is OK i will do that

Comment (mzitnik) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: @johnjcasey I have pushed the change testcontainers version

File: sdks/java/io/clickhouse/build.gradle [hunk 2]
```diff
@@ -49,7 +49,7 @@ idea {
   }
 }
 
-def clickhouse_jdbc_version = "0.2.6"
+def clickhouse_jdbc_version = "0.3.2-patch10"
```
Comment (mshustov) [inline sdks/java/io/clickhouse/build.gradle:52]: Why not to the latest version (`0.3.2-patch11`)?

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/TableSchema.java [hunk 3]
```diff
@@ -289,40 +289,42 @@ public static ColumnType parse(String str) {
...
-            return Long.valueOf(value);
-          default:
-            throw new UnsupportedOperationException("Unsupported type: " + columnType);
-        }
-      } catch (org.apache.beam.sdk.io.clickhouse.impl.parser.ParseException e) {
-        throw new IllegalArgumentException("failed to parse", e);
+      //      try {
+      String value = str;
+      /*
+      String value =
```
Comment (mshustov) [inline sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/TableSchema.java:295]: can be removed?

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java [hunk 4]
```diff
@@ -80,65 +58,21 @@ public static void setup() throws IOException, InterruptedException {
...
-        } else {
-          List<Image> images =
-              DockerClientFactory.instance().client().listImagesCmd().withShowAll(true).exec();
-          String listImagesOutput = "listImagesCmd:\n" + Joiner.on('\n').join(images) + "\n";
-
-          LOG.warn("failed to start clickhouse-server\n\n" + listImagesOutput, e);
-        }
-      }
-    }
+        new ClickHouseContainer("clickhouse/clickhouse-server:22.5")
```
Comment (mshustov) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java:61]: Why don't we test against the latest CH version?

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java [hunk 5]
```diff
@@ -80,65 +58,21 @@ public static void setup() throws IOException, InterruptedException {
...
+                BindMode.READ_ONLY);
+    ;
+    clickHouse.start();
+    LOG.info("Start Clickhouse");
   }
 
   @AfterClass
   public static void tearDown() {
     clickHouse.close();
-    zookeeper.close();
```
Comment (mshustov) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java:135]: Shouldn't we `close` it?

File: sdks/java/io/clickhouse/build.gradle (hunk 2 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/build.gradle:52]: Cause some conflict with testcontainers_version

Comment (mshustov) [reply sdks/java/io/clickhouse/build.gradle:52]: Sounds good

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java (hunk 4 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/BaseClickHouseTest.java]: Change to 22.9

**Comment by mzitnik** [comment]
@johnjcasey can you help with the PR

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java [hunk 6]
```diff
@@ -39,19 +41,19 @@ public class ClickHouseWriter {
...
-      ClickHouseRowBinaryStream stream, ColumnType columnType, Object value) throws IOException {
+  static void writeNullableValue(ClickHouseOutputStream stream, ColumnType columnType, Object value)
+      throws IOException {
 
     if (value == null) {
-      stream.markNextNullable(true);
+      BinaryStreamUtils.writeNull(stream);
     } else {
-      stream.markNextNullable(false);
+      BinaryStreamUtils.writeNonNull(stream);
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:50]: This api looks strange to me. Do you really need to manually write a null bit to the stream?

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java [hunk 7]
```diff
@@ -39,19 +41,19 @@ public class ClickHouseWriter {
...
       writeValue(stream, columnType, value);
     }
   }
 
   @SuppressWarnings("unchecked")
-  static void writeValue(ClickHouseRowBinaryStream stream, ColumnType columnType, Object value)
+  static void writeValue(ClickHouseOutputStream stream, ColumnType columnType, Object value)
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:56]: At this point, I believe value is known to be non-null, so we should be able to mark it as such and remove the suppressed warning

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java [hunk 8]
```diff
@@ -68,98 +70,98 @@ static void writeValue(ClickHouseRowBinaryStream stream, ColumnType columnType,
...
 
-  static void writeRow(ClickHouseRowBinaryStream stream, TableSchema schema, Row row)
+  static void writeRow(ClickHouseOutputStream stream, TableSchema schema, Row row)
       throws IOException {
     for (TableSchema.Column column : schema.columns()) {
       if (!column.materializedOrAlias()) {
         Object value = row.getValue(column.name());
 
         if (column.columnType().nullable()) {
+          // BinaryStreamUtils.writeNull(stream);
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:159]: remove this extra commented code please

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java [hunk 9]
```diff
@@ -56,7 +56,7 @@ private static boolean shouldAttempt(int i, long count) {
   /** With sufficient block size, ClickHouse will atomically insert all or nothing. */
   @Test
   public void testAtomicInsert() throws SQLException {
-    int size = 1000000;
+    int size = 100000;
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:59]: why did you shrink the size?

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java [hunk 10]
```diff
@@ -92,9 +92,10 @@ public void testAtomicInsert() throws SQLException {
    * With sufficient block size, ClickHouse will atomically insert all or nothing. In the case of
    * replicated tables, it will deduplicate blocks.
    */
+  //  @Test
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:95]: please remove this

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java [hunk 11]
```diff
@@ -92,9 +92,10 @@ public void testAtomicInsert() throws SQLException {
...
   @Test
   public void testIdempotentInsert() throws SQLException {
-    int size = 1000000;
+    int size = 100000;
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:98]: why is this smaller?

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java [hunk 12]
```diff
@@ -169,13 +169,17 @@ public void testParseArrayOfArrays() {
 
   @Test
   public void testParseDefaultExpressionString() {
-    assertEquals(
-        "abc", ColumnType.parseDefaultExpression(ColumnType.STRING, "CAST('abc' AS String)"));
+    assertEquals("abc", ColumnType.parseDefaultExpression(ColumnType.STRING, "abc"));
+    //    assertEquals(
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java:173]: please remove commented code

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java [hunk 13]
```diff
@@ -169,13 +169,17 @@ public void testParseArrayOfArrays() {
...
+    //            "abc", ColumnType.parseDefaultExpression(ColumnType.STRING, "CAST('abc' AS
+    // String)"));
+
   }
 
   @Test
   public void testParseDefaultExpressionInt64() {
-    assertEquals(-1L, ColumnType.parseDefaultExpression(ColumnType.INT64, "CAST(-1 AS Int64)"));
+    assertEquals(-1L, ColumnType.parseDefaultExpression(ColumnType.INT64, "-1"));
+    // assertEquals(-1L, ColumnType.parseDefaultExpression(ColumnType.INT64, "CAST(-1 AS Int64)"));
```
Comment (johnjcasey) [inline sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java:182]: same here

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java (hunk 6 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:50]: In ClickHouse Nullable mean that the field is optional in that case when  we do not have data we write null and when there is data we write not null and then the value

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java (hunk 7 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:56]: It is the same logic and i have only changed the API

File: sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java (hunk 8 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:159]: Removed

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java (hunk 9 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:59]: Fixing

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java (hunk 10 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:95]: Removed

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java (hunk 11 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:98]: Changed back

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java (hunk 13 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/TableSchemaTest.java:182]: Removed

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java (hunk 9 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:59]: setting 1M cause insert error. it also works with 500K rows, but i have changed it to 100K a deeper investigation need to be done here.

File: sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java (hunk 11 above)
Comment (mzitnik) [reply sdks/java/io/clickhouse/src/test/java/org/apache/beam/sdk/io/clickhouse/AtomicInsertTest.java:97]: setting 1M cause insert error. it also works with 500K rows, but i have changed it to 100K a deeper investigation need to be done here.

**Comment by damccorm** [comment]
stop reviewer notifications

**Comment by damccorm** [comment]
@johnjcasey has started this review and probably is the right person to complete it. I imagine responses may be slower since we are in the midst of the holiday season

**Review by johnjcasey — APPROVED** [review]

**Comment by johnjcasey** [comment]
run java precommit

**Comment by johnjcasey** [comment]
run java_pulsar_io_direct precommit

**Comment by mzitnik** [comment]
@johnjcasey do we know on what Apache Beam version it will be merged?

**Comment by johnjcasey** [comment]
it will be 2.45

**Comment by mshustov** [comment]
@johnjcasey is there ETA for v2.45 to communicate to our customers?

**Comment by johnjcasey** [comment]
The cut was last week, and we are closing out release blockers now. It could be as far away as 5 weeks from now, but more likely sooner

**Pull request merged**